scripts/jest/matchers/toWarnDev.js

```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+function normalizeCodeLocInfo(str) {
+  return str && str.replace(/at .+?:\d+/g, 'at **');
+}
+
+const createMatcherFor = consoleMethod =>
+  function matcher(callback, expectedMessages) {
+    if (__DEV__) {
+      // Warn about incorrect usage of matcher.
+      if (typeof expectedMessages === 'string') {
+        expectedMessages = [expectedMessages];
+      } else if (!Array.isArray(expectedMessages)) {
+        throw Error(
+          `toWarnDev() requires a parameter of type string or an array of strings ` +
+            `but was given ${typeof expectedMessages}.`
+        );
+      }
+
+      const consoleSpy = message => {
+        const normalizedMessage = normalizeCodeLocInfo(message);
+
+        for (let index = 0; index < expectedMessages.length; index++) {
+          const expectedMessage = expectedMessages[index];
+          if (
+            normalizedMessage === expectedMessage ||
+            normalizedMessage.includes(expectedMessage)
+          ) {
+            expectedMessages.splice(index, 1);
+            return;
+          }
+        }
+
+        // Fail early for unexpected warnings to preserve the call stack.
+        throw Error(`Unexpected warning recorded: "${message}"`);
```

Comment:
Can we include the *expected* warning message if one exists, in this message?

Otherwise it's a bit confusing that simply having a typo in the expected warning prints a message as if a warning was not expected at all.

Reply:
Yeah. I like this suggestion.

What do you think about the following change?
![screen shot 2018-01-02 at 10 50 01 am](https://user-images.githubusercontent.com/29597/34495593-b91371de-efaa-11e7-8dce-1196d37ea62f.png)

Reply:
To be clearer:
```js
it('should fail when the expected warnings passed to .toWarnDev() do not match the actual warnings', () => {
  expect(() => {
    console.error('foo');
  }).toWarnDev('bar');
});
```
![screen shot 2018-01-02 at 10 58 04 am](https://user-images.githubusercontent.com/29597/34495846-dae6bdc4-efab-11e7-9fd7-f654c294aa62.png)